File: kits/typescript/test/registry.test.ts
import { describe, it, expect, beforeEach, afterEach } from "vitest";
import { promises as fs } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { IronMcpRegistry, IronMcpEntry } from "../src/registry.js";

let dir: string;
beforeEach(async () => {
  dir = await fs.mkdtemp(join(tmpdir(), "ironmcp_reg_ts_"));
});
afterEach(async () => {
  await fs.rm(dir, { recursive: true, force: true });
});

describe("IronMcpRegistry", () => {
  it("concurrent registers do not lose an entry (lock closes TOCTOU, #9)", async () => {
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    await Promise.all([
      reg.register(new IronMcpEntry({ id: "a", namespace: "test", pid: 1 })),
      reg.register(new IronMcpEntry({ id: "b", namespace: "test", pid: 2 })),
      reg.register(new IronMcpEntry({ id: "c", namespace: "test", pid: 3 })),
      reg.register(new IronMcpEntry({ id: "d", namespace: "test", pid: 4 })),
    ]);
    const live = await reg.discover();
    expect(new Set(live.map((e) => e.id))).toEqual(new Set(["a", "b", "c", "d"]));
  });

  it("discover prunes a dead pid and rewrites the file (lazy GC, #10)", async () => {
    const reg = new IronMcpRegistry({ dir, isPidAlive: (pid) => pid !== 2 });
    await reg.register(new IronMcpEntry({ id: "a", namespace: "test", pid: 1 }));
    await reg.register(new IronMcpEntry({ id: "b", namespace: "test", pid: 2 })); // dead
    expect(new Set((await reg.discover()).map((e) => e.id))).toEqual(new Set(["a"]));
    // rewritten: a second discover still only sees the live one
    expect(new Set((await reg.discover()).map((e) => e.id))).toEqual(new Set(["a"]));
    // and the on-disk file no longer holds the dead entry
    const onDisk = JSON.parse(await fs.readFile(join(dir, "registry.json"), "utf8"));
    expect(Object.keys(onDisk)).toEqual(["a"]);
  });

  it("unregister removes an entry", async () => {
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    await reg.register(new IronMcpEntry({ id: "a", namespace: "test", pid: 1 }));
    await reg.unregister("a");
    expect(await reg.discover()).toEqual([]);
  });

  it("entry JSON is language-neutral and carries NO hand-kept tool list (#3)", () => {
    const j = new IronMcpEntry({
      id: "x",
      namespace: "ns",
      pid: 9,
      host: "127.0.0.1",
      port: 8080,
      transport: "http",
      version: "1.0",
      codeSha: "abc123",
      capabilities: { tools: {} },
    }).toJSON();
    expect(j.code_sha).toBe("abc123");
    expect(typeof j.started_at).toBe("string");
    expect(j).not.toHaveProperty("tools"); // honesty: no drifting count
    const e = IronMcpEntry.fromJSON(j);
    expect(e.id).toBe("x");
    expect(e.port).toBe(8080);
    expect(e.transport).toBe("http");
  });

  it("started_at is ISO-8601 UTC and the file is flat-keyed-by-id (Dart-compatible fabric)", async () => {
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    await reg.register(new IronMcpEntry({ id: "srv1", namespace: "ns", pid: 1, port: 9000 }));
    const onDisk = JSON.parse(await fs.readFile(join(dir, "registry.json"), "utf8"));
    // top level keyed by entry id, matching Dart `map[entry.id] = entry.toJson()`
    expect(Object.keys(onDisk)).toEqual(["srv1"]);
    expect(onDisk.srv1.namespace).toBe("ns");
    // ISO-8601 UTC (trailing Z), parseable back to a Date
    expect(onDisk.srv1.started_at).toMatch(/T.*Z$/);
    expect(Number.isNaN(new Date(onDisk.srv1.started_at).getTime())).toBe(false);
  });

  it("started_at is the CANONICAL registry format: ISO-8601 UTC, millisecond precision, trailing Z", async () => {
    // The estate canon is exactly YYYY-MM-DDTHH:mm:ss.sssZ (3 fractional digits, Z — never a
    // +00:00 offset, never 6-digit microseconds). Pin it so a future refactor can't drift the
    // registry.json off byte-parity with the other kits once they normalize to this shape.
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    await reg.register(new IronMcpEntry({ id: "srv1", namespace: "ns", pid: 1 }));
    const onDisk = JSON.parse(await fs.readFile(join(dir, "registry.json"), "utf8"));
    expect(onDisk.srv1.started_at).toMatch(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/);
    expect(onDisk.srv1.started_at).not.toContain("+00:00"); // not the Python-style offset
    expect(onDisk.srv1.started_at).not.toMatch(/\.\d{6}Z$/); // not Dart-style microseconds

    // A fixed instant round-trips to the exact canonical string, byte-for-byte.
    const fixed = new Date("2026-09-01T10:35:34.123Z");
    const j = new IronMcpEntry({ id: "x", namespace: "ns", pid: 9, startedAt: fixed }).toJSON();
    expect(j.started_at).toBe("2026-09-01T10:35:34.123Z");
  });

  it("discover's prune is PERSISTED: a FRESH registry instance re-reading disk sees the dead entry gone (#10)", async () => {
    // Not just an in-memory re-prune — write the dead entry, prune via one instance, then open a
    // BRAND-NEW registry (that always reports pids alive) and confirm the dead entry is truly gone
    // from registry.json on disk.
    const writer = new IronMcpRegistry({ dir, isPidAlive: () => true });
    await writer.register(new IronMcpEntry({ id: "live", namespace: "t", pid: 1 }));
    await writer.register(new IronMcpEntry({ id: "dead", namespace: "t", pid: 2 }));
    const pruner = new IronMcpRegistry({ dir, isPidAlive: (pid) => pid !== 2 });
    await pruner.discover(); // prunes "dead" and must rewrite the file
    // A fresh reader that considers EVERY pid alive would still surface "dead" if the rewrite was
    // skipped. It does not -> the prune reached disk.
    const fresh = new IronMcpRegistry({ dir, isPidAlive: () => true });
    expect(new Set((await fresh.discover()).map((e) => e.id))).toEqual(new Set(["live"]));
    const onDisk = JSON.parse(await fs.readFile(join(dir, "registry.json"), "utf8"));
    expect(Object.keys(onDisk)).toEqual(["live"]);
  });

  it("a corrupt registry.json starts fresh rather than crashing", async () => {
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(join(dir, "registry.json"), "{ this is not json ]]", "utf8");
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    // discover does not throw; it treats the garbage as empty
    expect(await reg.discover()).toEqual([]);
    // and register still lands over the top of the corruption
    await reg.register(new IronMcpEntry({ id: "a", namespace: "t", pid: 1 }));
    expect((await reg.discover()).map((e) => e.id)).toEqual(["a"]);
  });

  it("a whitespace-only registry.json is treated as empty (the trim -> {} branch)", async () => {
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(join(dir, "registry.json"), "   \n\t  ", "utf8");
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    expect(await reg.discover()).toEqual([]);
    await reg.register(new IronMcpEntry({ id: "b", namespace: "t", pid: 1 }));
    expect((await reg.discover()).map((e) => e.id)).toEqual(["b"]);
  });

  it("a non-object top-level (a JSON array) starts fresh, does not iterate garbage keys", async () => {
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(join(dir, "registry.json"), "[1, 2, 3]", "utf8");
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    expect(await reg.discover()).toEqual([]);
  });

  it("register on a corrupt file recovers (register's read() also fails open)", async () => {
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(join(dir, "registry.json"), "\0\0not-json", "utf8");
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true });
    await reg.register(new IronMcpEntry({ id: "z", namespace: "t", pid: 1 }));
    const onDisk = JSON.parse(await fs.readFile(join(dir, "registry.json"), "utf8"));
    expect(Object.keys(onDisk)).toEqual(["z"]);
  });

  it("a LIVE lock held past lockTimeoutMs is given up on (proceed best-effort, NOT stolen)", async () => {
    // A fresh (non-stale) lock that a still-live holder keeps: the deadline must break the wait and
    // proceed best-effort, rather than stealing the lock (that only happens past staleLockAfterMs).
    const lock = join(dir, "registry.json.lock");
    await fs.mkdir(dir, { recursive: true });
    await fs.writeFile(lock, ""); // mtime = now, well within staleLockAfter
    const reg = new IronMcpRegistry({
      dir,
      isPidAlive: () => true,
      lockTimeoutMs: 40, // give up quickly
      staleLockAfterMs: 60_000, // never treat this lock as stale
    });
    const t0 = Date.now();
    await reg.register(new IronMcpEntry({ id: "a", namespace: "t", pid: 1 }));
    // it proceeded (write landed) after waiting ~lockTimeoutMs, and did NOT remove the live lock
    expect((await reg.discover()).map((e) => e.id)).toEqual(["a"]);
    expect(Date.now() - t0).toBeGreaterThanOrEqual(30);
    await expect(fs.stat(lock)).resolves.toBeTruthy(); // the live holder's lock is left intact
  });

  it("the DEFAULT pid-liveness (real process.kill) sees this process alive and a bogus pid dead", async () => {
    // No injected isPidAlive -> exercises pidAliveDefault. Our own pid is alive; a huge unused pid
    // yields ESRCH -> dead and is pruned.
    const reg = new IronMcpRegistry({ dir }); // real pidAliveDefault
    await reg.register(new IronMcpEntry({ id: "self", namespace: "t", pid: process.pid }));
    await reg.register(new IronMcpEntry({ id: "ghost", namespace: "t", pid: 2_000_000_000 }));
    expect(new Set((await reg.discover()).map((e) => e.id))).toEqual(new Set(["self"]));
  });

  it("fromJSON READS cross-kit timestamps robustly (Dart microseconds, Python +00:00, unparseable)", () => {
    // Dart-style microsecond precision — parseable by Date, kept.
    const dartMicro = IronMcpEntry.fromJSON({
      id: "d",
      namespace: "t",
      pid: 1,
      started_at: "2026-09-01T10:35:34.123456Z",
    });
    expect(Number.isNaN(dartMicro.startedAt.getTime())).toBe(false);

    // Python-style +00:00 offset — parseable by Date, kept.
    const pyOffset = IronMcpEntry.fromJSON({
      id: "p",
      namespace: "t",
      pid: 1,
      started_at: "2026-09-01T10:35:34.123456+00:00",
    });
    expect(Number.isNaN(pyOffset.startedAt.getTime())).toBe(false);

    // Unparseable string -> the Number.isNaN fallback to new Date() (a valid instant, never NaN).
    const junk = IronMcpEntry.fromJSON({ id: "j", namespace: "t", pid: 1, started_at: "not-a-date" });
    expect(Number.isNaN(junk.startedAt.getTime())).toBe(false);

    // Non-string started_at (a number) -> typeof guard falls to new Date(), still valid.
    const nonString = IronMcpEntry.fromJSON({ id: "n", namespace: "t", pid: 1, started_at: 12345 as unknown as string });
    expect(Number.isNaN(nonString.startedAt.getTime())).toBe(false);
  });

  it("a stale lock left by a crashed holder is stolen so writes still land", async () => {
    // Pre-create the lock and back-date it well past staleLockAfter.
    const lock = join(dir, "registry.json.lock");
    await fs.writeFile(lock, "");
    const old = new Date(Date.now() - 60_000);
    await fs.utimes(lock, old, old);
    const reg = new IronMcpRegistry({ dir, isPidAlive: () => true, staleLockAfterMs: 30_000 });
    await reg.register(new IronMcpEntry({ id: "a", namespace: "test", pid: 1 }));
    expect((await reg.discover()).map((e) => e.id)).toEqual(["a"]);
  });
});

File: kits/dart/test/registry_test.dart
import 'dart:io';

import 'package:ironmcp/ironmcp.dart';
import 'package:test/test.dart';

void main() {
  late Directory dir;
  setUp(() async => dir = await Directory.systemTemp.createTemp('ironmcp_reg'));
  tearDown(() async => dir.delete(recursive: true));

  test('concurrent registers do not lose an entry (lock closes TOCTOU, #9)',
      () async {
    final reg = IronMcpRegistry(dir: dir, isPidAlive: (_) => true);
    await Future.wait([
      reg.register(IronMcpEntry(id: 'a', namespace: 'test', pid: 1)),
      reg.register(IronMcpEntry(id: 'b', namespace: 'test', pid: 2)),
      reg.register(IronMcpEntry(id: 'c', namespace: 'test', pid: 3)),
      reg.register(IronMcpEntry(id: 'd', namespace: 'test', pid: 4)),
    ]);
    final live = await reg.discover();
    expect(live.map((e) => e.id).toSet(), {'a', 'b', 'c', 'd'});
  });

  test('discover prunes a dead pid and rewrites the file (lazy GC, #10)',
      () async {
    final reg = IronMcpRegistry(dir: dir, isPidAlive: (pid) => pid != 2);
    await reg.register(IronMcpEntry(id: 'a', namespace: 'test', pid: 1));
    await reg.register(IronMcpEntry(id: 'b', namespace: 'test', pid: 2)); // dead
    expect((await reg.discover()).map((e) => e.id).toSet(), {'a'});
    // rewritten: a second discover still only sees the live one
    expect((await reg.discover()).map((e) => e.id).toSet(), {'a'});
  });

  test('unregister removes an entry', () async {
    final reg = IronMcpRegistry(dir: dir, isPidAlive: (_) => true);
    await reg.register(IronMcpEntry(id: 'a', namespace: 'test', pid: 1));
    await reg.unregister('a');
    expect(await reg.discover(), isEmpty);
  });

  test('entry JSON is language-neutral and carries NO hand-kept tool list (#3)',
      () {
    final j = IronMcpEntry(
      id: 'x',
      namespace: 'ns',
      pid: 9,
      host: '127.0.0.1',
      port: 8080,
      transport: 'http',
      version: '1.0',
      codeSha: 'abc123',
      capabilities: {'tools': <String, dynamic>{}},
    ).toJson();
    expect(j['code_sha'], 'abc123');
    expect(j['started_at'], isA<String>());
    expect(j.containsKey('tools'), isFalse); // honesty: no drifting count
    final e = IronMcpEntry.fromJson(j);
    expect(e.id, 'x');
    expect(e.port, 8080);
    expect(e.transport, 'http');
  });

  // TASK 2 — the CANONICAL registry started_at format. Every ironmcp kit MUST
  // emit this exact shape so registry.json is byte-identical across languages:
  // ISO-8601 UTC, MILLISECOND precision (exactly 3 fractional digits), trailing
  // Z — e.g. 2026-09-01T10:35:34.123Z. NOT a +00:00 offset; NOT 6-digit micros.
  final canonical = RegExp(r'^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$');

  test('isoMillisUtc emits exactly 3 fractional digits + Z (no offset)', () {
    // A microsecond-bearing instant is where Dart\'s own toIso8601String() would
    // blow out to 6 digits — the exact drift this normalisation fixes.
    final dt = DateTime.utc(2026, 9, 1, 10, 35, 34, 123, 456);
    expect(isoMillisUtc(dt), '2026-09-01T10:35:34.123Z');
    expect(canonical.hasMatch(isoMillisUtc(dt)), isTrue);
    // Micros are TRUNCATED (floored), never rounded: .123456 -> .123, .999999 -> .999.
    expect(isoMillisUtc(DateTime.utc(2026, 9, 1, 0, 0, 0, 999, 999)),
        '2026-09-01T00:00:00.999Z');
  });

  test('isoMillisUtc zero-pads every field and converts local to UTC', () {
    expect(isoMillisUtc(DateTime.utc(1, 2, 3, 4, 5, 6, 7)),
        '0001-02-03T04:05:06.007Z');
    // A non-UTC input is normalised to UTC before formatting.
    final local = DateTime.utc(2026, 1, 1, 12, 0, 0).toLocal();
    expect(isoMillisUtc(local), '2026-01-01T12:00:00.000Z');
  });

  test('entry started_at is the canonical millisecond-Z string, no +00:00', () {
    final j = IronMcpEntry(
      id: 'x',
      namespace: 'ns',
      pid: 9,
      startedAt: DateTime.utc(2026, 9, 1, 10, 35, 34, 123, 456),
    ).toJson();
    expect(j['started_at'], '2026-09-01T10:35:34.123Z');
    expect((j['started_at'] as String).contains('+00:00'), isFalse);
  });

  test('a default (now) entry still matches the canonical shape', () {
    final j = IronMcpEntry(id: 'x', namespace: 'ns', pid: 9).toJson();
    expect(canonical.hasMatch(j['started_at'] as String), isTrue,
        reason: 'default started_at must be YYYY-MM-DDTHH:mm:ss.SSSZ');
  });

  test('started_at round-trips: fromJson(toJson) re-emits the same string', () {
    final entry = IronMcpEntry(
      id: 'x',
      namespace: 'ns',
      pid: 9,
      startedAt: DateTime.utc(2026, 9, 1, 10, 35, 34, 123, 456),
    );
    final s1 = entry.toJson()['started_at'];
    final s2 = IronMcpEntry.fromJson(entry.toJson()).toJson()['started_at'];
    expect(s1, '2026-09-01T10:35:34.123Z');
    expect(s2, s1); // parse then re-emit is a fixed point
  });
}
